File: illustration/licensing.py
"""Licence-code normalisation — one vocabulary for four provider vocabularies.

Every provider spells the same licence differently: Openverse emits ``by-sa``,
Wikimedia Commons emits ``cc-by-sa-4.0`` (and ``CC BY-SA 4.0`` as the short
name), Pixabay emits ``Pixabay License``. :func:`normalize_license` folds those
spellings onto one canonical code so the licence gate
(:func:`illustration.schema.license_allowlist`) can compare like with like,
instead of silently returning ``[]`` for two of the four built-in sources.

**The safety invariant, and it is the whole reason this module is small and
boring: normalisation may never drop a restriction token.** ``nc``
(non-commercial) and ``nd`` (no-derivatives) are exactly what an allowlist
exists to exclude, so the transform only ever removes things that carry no
permission meaning — a ``cc``/``cc-`` prefix and a trailing version number —
and every other spelling must be listed, by hand, in :data:`LICENSE_ALIASES`.
A code that is not recognised stays as-is and therefore fails the allowlist:
**unknown is not allowed**, in both directions.

>>> normalize_license("cc-by-sa-4.0")
'by-sa'
>>> normalize_license("CC BY-SA 4.0")
'by-sa'
>>> normalize_license("Pixabay License")
'pixabay-license'
>>> normalize_license("cc-by-nc-nd-4.0")          # restrictions survive intact
'by-nc-nd'
>>> normalize_license(None) is None
True
"""

from __future__ import annotations

import re

__all__ = ["normalize_license", "LICENSE_ALIASES", "RESTRICTION_TOKENS"]

#: Whole-code spellings that no mechanical rule can fold, mapped by hand. Kept
#: deliberately short: every entry is a judgement that two strings name the same
#: permissions, and a wrong entry silently widens the gate.
LICENSE_ALIASES = {
    "cc-zero": "cc0",
    "cc-0": "cc0",
    "zero": "cc0",
    "publicdomain": "pdm",
    "public-domain": "pdm",
    "public-domain-mark": "pdm",
    "pd": "pdm",
    "pdm-owner": "pdm",
    "cc-pdm": "pdm",
    "cc-publicdomain": "pdm",
}

#: Tokens that *restrict* use. Present only so the guard test can assert that
#: normalisation never removes one; nothing in the transform consults it.
RESTRICTION_TOKENS = frozenset({"nc", "nd", "sampling"})

# A trailing version, e.g. the "-4.0" of "cc-by-sa-4.0". The leading separator is
# required so a code that merely *ends* in a digit (``cc0``) is left alone.
_VERSION_SUFFIX_RE = re.compile(r"[-_ ]v?\d+(?:\.\d+)*$")
# The Creative Commons prefix, which carries no permission meaning of its own.
_CC_PREFIX_RE = re.compile(r"^cc[-_ ]")
_SEPARATORS_RE = re.compile(r"[\s_]+")


def normalize_license(value: "str | None") -> "str | None":
    """Fold a provider's licence spelling onto one canonical, comparable code.

    Lower-cases, unifies separators to ``-``, strips a trailing version and a
    leading ``cc-``, then applies :data:`LICENSE_ALIASES`. Returns ``None`` for
    ``None``/blank — an absent licence is never a code.

    >>> normalize_license("by-sa"), normalize_license("cc0"), normalize_license("CC0 1.0")
    ('by-sa', 'cc0', 'cc0')
    >>> normalize_license("Pexels License"), normalize_license("  BY  ")
    ('pexels-license', 'by')
    >>> normalize_license("public domain"), normalize_license("cc-by-3.0")
    ('pdm', 'by')
    >>> normalize_license("")  is None
    True
    """
    if not value or not value.strip():
        return None
    code = _SEPARATORS_RE.sub("-", value.strip().lower())
    code = _VERSION_SUFFIX_RE.sub("", code)
    if code in LICENSE_ALIASES:  # alias before de-prefixing (e.g. "cc-pdm")
        return LICENSE_ALIASES[code]
    code = _CC_PREFIX_RE.sub("", code)
    return LICENSE_ALIASES.get(code, code) or None

File: ts/src/licensing.ts
/**
 * Licence normalisation and the allowlist gate, mirroring `illustration.licensing`
 * and `illustration.schema.license_allowlist`.
 *
 * The alias table is generated data, not re-typed here. The transform is the
 * same three steps in the same order as Python, because the order is what makes
 * `cc-0` reachable (the version strip would otherwise eat its `-0`), and the
 * invariant is the same: **a restriction token is never dropped**. The Python
 * side's `license_normalization_cases` are replayed in `parity.test.ts`.
 */

import { CONSTANTS } from './generated/constants';
import type { ImageResult } from './generated/image-result';

const ALIASES: Readonly<Record<string, string>> = CONSTANTS.license_aliases;

/** The default commercial-safe allowlist (`DFLT_LICENSE_ALLOWLIST` on the Python side). */
export const DEFAULT_LICENSE_ALLOWLIST: readonly string[] = CONSTANTS.default_license_allowlist;

// A trailing version, e.g. the "-4.0" of "cc-by-sa-4.0"; the leading separator is
// required so a code that merely ends in a digit (`cc0`) is left alone.
const VERSION_SUFFIX = /[-_ ]v?\d+(?:\.\d+)*$/;
const CC_PREFIX = /^cc[-_ ]/;
const SEPARATORS = /[\s_]+/g;

/** Fold a provider's licence spelling onto one canonical, comparable code.
 *  `null` for null/blank: an absent licence is never a code. */
export function normalizeLicense(value: string | null | undefined): string | null {
  if (!value || !value.trim()) return null;
  let code = value.trim().toLowerCase().replace(SEPARATORS, '-');
  if (Object.hasOwn(ALIASES, code)) return ALIASES[code]!;
  code = code.replace(VERSION_SUFFIX, '');
  if (Object.hasOwn(ALIASES, code)) return ALIASES[code]!;
  code = code.replace(CC_PREFIX, '');
  return (Object.hasOwn(ALIASES, code) ? ALIASES[code]! : code) || null;
}

/** Keep only results whose licence is on the allowlist; unknown is not allowed.
 *  Both sides are normalised, so provider dialects match without enumeration. */
export function licenseAllowlist<T extends Pick<ImageResult, 'license'>>(
  results: Iterable<T>,
  opts: { allow?: Iterable<string> | null } = {},
): T[] {
  const allowed = new Set<string>();
  for (const code of opts.allow ?? DEFAULT_LICENSE_ALLOWLIST) {
    const normalized = normalizeLicense(code);
    if (normalized !== null) allowed.add(normalized);
  }
  const kept: T[] = [];
  for (const result of results) {
    const code = normalizeLicense(result.license);
    if (code !== null && allowed.has(code)) kept.push(result);
  }
  return kept;
}
